bob-backend/src/core.rs

```diff
@@ -131,10 +134,74 @@ pub trait MetricsProducer: Send + Sync {
     }
 }
 
+#[derive(Hash, PartialEq, Eq, Debug)]
+enum BackendErrorAction {
+    PUT(String, String),
+}
+
+#[derive(Debug)]
+struct BackendErrorLogger {
+    errors: HashMap<BackendErrorAction, u64>,
+    interval: Duration
+}
+
+impl BackendErrorLogger {
+    fn new(interval: Duration) -> BackendErrorLogger {
+        BackendErrorLogger {
+            errors: HashMap::new(),
+            interval,
+        }
+    }
+
+    fn report_error(&mut self, action: BackendErrorAction) {
+        if let Some(count) = self.errors.get_mut(&action) {
+            *count += 1;
+        } else {
+            self.errors.insert(action, 1);
```

Comment:
This is not thread safe. `task` function can access `errors` while this `insert` in progress